python/ray/train/v2/_internal/state/state_actor.py

```diff
@@ -1,47 +1,165 @@
+        self, last_poll_run_id: Optional[str]
+    ) -> str:
+        aborted_run_ids = []
+        with self._runs_lock:
+            runs = list(self._runs.values())
+
+            # Start iterating from poll index.
+            poll_index = 0
+            if last_poll_run_id is not None:
+                for i, run in enumerate(runs):
+                    if run.id == last_poll_run_id:
+                        poll_index = (i + 1) % len(runs)
+                        break
+
+            # Abort runs.
+            num_sampled_runs = 0
+            i = poll_index
+            while (
+                i < poll_index + len(runs)
+                and num_sampled_runs < self._controllers_to_poll_per_iteration
+            ):
+                run = runs[i % len(runs)]
+                i += 1
+                last_poll_run_id = run.id
+                if run.status.is_terminal():
+                    continue
+                actor_state = get_actor(
+                    run.controller_actor_id, timeout=self._get_actor_timeout_s
+                )
+                num_sampled_runs += 1
+                if not actor_state or actor_state.state == "DEAD":
+                    update_train_run_aborted(run, False)
+                    self.create_or_update_train_run(run)
+                    aborted_run_ids.append(run.id)
+
+        # Abort run attempts.
+        with self._run_attempts_lock:
+            for run_id in aborted_run_ids:
+                latest_run_attempt = self._get_latest_run_attempt(run_id)
+                if latest_run_attempt and not latest_run_attempt.status.is_terminal():
+                    update_train_run_attempt_aborted(latest_run_attempt, False)
+                    self.create_or_update_train_run_attempt(latest_run_attempt)
+
+        return last_poll_run_id
+
+    def _start_run_state_reconciliation_thread(self) -> None:
+        def _reconciliation_loop():
+            last_poll_run_id = None
+            latest_poll_time = float("-inf")
+            while True:
+                # Wait for the poll interval to elapse.
+                time_since_last_poll = time_monotonic() - latest_poll_time
+                if time_since_last_poll < self._poll_interval_s:
+                    remaining_time = self._poll_interval_s - time_since_last_poll
+                    time.sleep(remaining_time)
+
+                last_poll_run_id = self._abort_live_runs_with_dead_controllers(
+                    last_poll_run_id
+                )
+                latest_poll_time = time_monotonic()
+
+        thread = threading.Thread(target=_reconciliation_loop, daemon=True)
+        thread.start()
+
+    def _get_latest_run_attempt(self, run_id: str) -> Optional[TrainRunAttempt]:
+        with self._run_attempts_lock:
+            run_attempts = self._run_attempts.get(run_id, {})
+            if not run_attempts:
+                return None
+            return next(reversed(run_attempts.values()))
```

Comment:
how is the run attempts dict ordered?

Reply:
Added this comment: `# NOTE: run_attempts is OrderedDict from attempt_id to TrainRunAttempt.` Lmk if this makes sense.